File: leetcode/1971. Find if Path Exists in Graph/main.go
// SOlved by @kitanoyoru
package main

func contains(arr []int, val int) bool {
	for _, v := range arr {
		if v == val {
			return true
		}
	}

	return false
}

func validPath(n int, edges [][]int, source int, destination int) bool {
	if len(edges) == 0 && source == destination {
		return true
	}

	visited := make([]bool, n)
	graph := make(map[int][]int)
	ans := false

	var dfs func(int, int)
	dfs = func(start, end int) {
		if !visited[start] && !ans {
			if start == end {
				ans = true
				return
			}

			visited[start] = true
			for _, neigh := range graph[start] {
				dfs(neigh, end)
			}
		}
	}

	for _, edge := range edges {
		if !contains(graph[edge[0]], edge[1]) {
			graph[edge[0]] = append(graph[edge[0]], edge[1])
		}
		if !contains(graph[edge[1]], edge[0]) {
			graph[edge[1]] = append(graph[edge[1]], edge[0])
		}
	}

	if contains(graph[source], destination) {
		return true
	}

	dfs(source, destination)

	return ans
}

File: leetcode/1971. Find if Path Exists in Graph/main.swift
typealias Graph = [Int: [Int]]

class Solution {
    func validPath(_ n: Int, _ edges: [[Int]], _ source: Int, _ destination: Int) -> Bool {
        let graph = buildGraph(edges: edges)
        return hasPath(graph, source: source, destination: destination)
    }

    private func buildGraph(edges: [[Int]]) -> Graph {
        var graph: Graph = [:]
        
        for edge in edges {
            let node1 = edge[0], node2 = edge[1]
            graph[node1, default: []].append(node2)
            graph[node2, default: []].append(node1)
        }
        
        return graph
    }

    private func hasPath(_ graph: Graph, source: Int, destination: Int) -> Bool {
        var visited: Set<Int> = []
        return dfs(graph, current: source, target: destination, visited: &visited)
    }

    private func dfs(_ graph: Graph, current: Int, target: Int, visited: inout Set<Int>) -> Bool {
        if current == target { return true }
        if visited.contains(current) { return false }
        
        visited.insert(current)
        for neighbor in graph[current] ?? [] {
            if dfs(graph, current: neighbor, target: target, visited: &visited) {
                return true
            }
        }
        
        return false
    }
}

